infracloudio/kyverno
# issue-5100:added logic to check nsPolicies

Signed-off-by: SANSKARJAIN2 <sanskar.jain@infracloud.io>

## Explanation
Upon editing the generated resource, only cluster policies are fetched to synchronize the resource with the source. This PR contains logic to support sync when the resource is generated via a namespaced policy.
 


## Related issue
closes: 5100


## Milestone of this PR


## What type of PR is this



## Proposed Changes
In order to sync the generated resource with the source resource when the generated resource is edited, only the cluster policies were fetched.
To detect which policy to fetch( cluster or namespaced) a new label is introduced `policy.kyverno.io/policy-kind` which can have can have either `Namespace` or `Cluster` indicating the kind of policy that created this resource.
The default behavior is to fetch cluster policies.


### Proof Manifests
1. Create a namespace `poltest`
`kubectl create ns poltest`
2. Create secret named regcred
`kubectl -n poltest create secret generic regcred --from-literal=foo=bar`
3. Create policy
```
apiVersion: kyverno.io/v2beta1
kind: Policy
metadata:
  name: pol-sync-clone
  namespace: poltest
spec:
  rules:
  - name: gen-zk
    match:
      any:
      - resources:
          kinds:
          - ConfigMap
    generate:
      apiVersion: v1
      kind: Secret
      name: myclonedsecret
      namespace: poltest
      synchronize: true
      clone:
        namespace: poltest
        name: regcred 
```
4. Create a configMap in poltest namespace
`k -n poltest create cm foo --from-literal=foo=bar`
5. Verify that secret `myclonedsecret` is created in poltest namespace have same value as regcred in poltest namespace 
6.  edit `myclonedsecret` and wait for 2-3 seconds.
7. Verify that `myclonedsecret` is reverted to the original value ( same as `regcred`)



## Checklist



  
  
  
  

## Further Comments
